crates/ruff_python_semantic/src/analyze/visibility.rs

```diff
@@ -138,24 +125,38 @@ fn stem(path: &str) -> &str {
     }
 }
 
+/// A Python module can either be defined as a module path (i.e., the dot-separated path to the
+/// module) or, if the module can't be resolved, as a file path (i.e., the path to the file defining
+/// the module).
+#[derive(Debug)]
+pub enum ModuleSource<'a> {
+    /// A module path is a dot-separated path to the module.
+    Path(&'a [String]),
+    /// A file path is the path to the file defining the module, often a script outside of a
+    /// package.
+    File(&'a Path),
+}
+
 /// Return the `Visibility` of the Python file at `Path` based on its name.
-pub fn module_visibility(module_path: Option<&[String]>, path: &Path) -> Visibility {
-    if let Some(module_path) = module_path {
-        if module_path.iter().any(|m| is_private_module(m)) {
-            return Visibility::Private;
-        }
-    } else {
-        // When module_path is None, path is a script outside a package, so just
-        // check to see if the module name itself is private.
-        // Ex) `_foo.py` (but not `__init__.py`)
-        let mut components = path.iter().rev();
-        if let Some(filename) = components.next() {
-            let module_name = filename.to_string_lossy();
-            let module_name = stem(&module_name);
-            if is_private_module(module_name) {
+pub fn module_visibility(source: &ModuleSource) -> Visibility {
```

Comment:
Nit: Either implement as `to_visibility(&self)` on `ModuleSource` or as `Visibility::from_module_source(source: &Modulesource)`. Adding it to an object rather than having free-floating functions can help with discoverability.